### PR TITLE
Test embedding with gat

### PR DESCRIPTION
**Pull request by jpngfile**
I fixed and refactored the preprocessing code, and added a new model that combines the embedding with the GAT layer. I still need to test this model, so I added some empty functions marking future work.

## Preprocessing changes
- Renamed `dfs` to `bfs` to be more accurate
- Refactored loading GloVe embedding to a separate file
- Removed `ngrams_iterator` since it's redundant in this case
- Renamed `create_adj_matrix` to `assign_indices_to_nodes`
- Fixed `get_adj_matrix` to return the sparse tensor
- Added code to save processed data of a single tree to a pkl file
- Changed output to return a float tensor, instead of a boolean tensor
- Changed `get_adj_matrix` so a node is its own neighbour. I realized that GAT doesn't have a way to process nodes with no neighbours, so by the paper, a node is always its own neighbour. We may change this if we can find a way to deal with neighbourless nodes.

## Model Changes
- Changed GAT model so it accepts the adjacency matrix in its `forward` function. In order to be able to predict on different graph structures, the adjacency matrix should be an input, rather than passed in initialization.
- Added the `GloveGAT` model. It combines an `EmbeddingBag` layer to a `GATFInal` layer. The embedding bag will average the word vectors of a sentence to pass the the GAT layer.

So far I have just created a scaffold of the new model and is subject to testing and changes. I still need to:
- Create functions for metrics (i.e. precision and recall)
- Create functions to split data into training and test
- Create functions to combine multiple trees together

### CODE REVIEW

File: process_trees.py [hunk 1]
```diff
@@ -87,8 +69,16 @@ def get_output(trees):
...
     embedding, offsets = get_tree_text_embedding(texts)
-
+    outputs = get_output(tree)
+
+    model_data = {
+        "adj_matrix": adj_mat,
+        "embedding": embedding,
+        "offsets": offsets,
+        "outputs": outputs
+    }
```
Comment (agpar) [inline process_trees.py:83]: So `tree_indices` is not necessary context here because everything has been translated to ordered lists, and the `i`th element in each of the `model_data` elements corresponds to the same node, right?

Comment (jpngfile) [reply process_trees.py:83]: Yes, tree_indices is just a way to order the nodes to create the adjacency matrix.
